overlord/state/warning.go

```diff
@@ -0,0 +1,246 @@
+	lastSeen time.Time
+	// the last time one of these was shown to the user
+	lastShown time.Time
+	// how much time since one of these was last seen should we drop the message
+	deleteAfter time.Duration
+	// how much time since one of these was last shown should we repeat it
+	repeatAfter time.Duration
+}
+
+func (w *Warning) String() string {
+	return w.message
+}
+
+func (w *Warning) MarshalJSON() ([]byte, error) {
+	jw := jsonWarning{
+		Message:     w.message,
+		FirstSeen:   w.firstSeen,
+		LastSeen:    w.lastSeen,
+		DeleteAfter: w.deleteAfter.String(),
+		RepeatAfter: w.repeatAfter.String(),
+	}
+	if !w.lastShown.IsZero() {
+		jw.LastShown = &w.lastShown
+	}
+
+	return json.Marshal(jw)
+}
+
+func (w *Warning) UnmarshalJSON(data []byte) error {
+	var jw jsonWarning
+	err := json.Unmarshal(data, &jw)
+	if err != nil {
+		return err
+	}
+	w.message = jw.Message
+	w.firstSeen = jw.FirstSeen
+	w.lastSeen = jw.LastSeen
+	if jw.LastShown != nil {
+		w.lastShown = *jw.LastShown
+	}
+	w.deleteAfter, err = time.ParseDuration(jw.DeleteAfter)
+	if err != nil {
+		return err
+	}
+	w.repeatAfter, err = time.ParseDuration(jw.RepeatAfter)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (w *Warning) IsDeletable(now time.Time) bool {
+	return w.lastSeen.Add(w.deleteAfter).Before(now)
+}
+
+func (w *Warning) IsShowable(t time.Time) bool {
+	return (w.lastShown.IsZero() || w.lastShown.Add(w.repeatAfter).Before(t)) && !w.firstSeen.After(t)
+}
+
+// flattenWarning loops over the warnings map, and returns all
+// warnings therein as a flat list, for serialising.
+// Call with the lock held.
+func (s *State) flattenWarnings() []*Warning {
+	flat := make([]*Warning, 0, len(s.warnings))
+	for _, w := range s.warnings {
+		flat = append(flat, w)
+	}
+	return flat
+}
+
+// unflattenWarnings takes a flat list of warnings and replaces the
+// warning map with them.
+// Call with the lock held.
+func (s *State) unflattenWarnings(flat []*Warning) {
+	s.warnings = make(map[string]*Warning, len(flat))
+	for _, w := range flat {
+		s.warnings[w.message] = w
+	}
+}
+
+// AddWarning records a warning: if it's the first Warning with this
+// message it'll be added (with its firstSeen and lastSeen set to the
+// current time), otherwise the existing one will have its lastSeen
+// updated.
+func (s *State) AddWarning(message string) {
+	s.addWarningFull(Warning{
+		message:     message,
+		deleteAfter: DefaultDeleteAfter,
+		repeatAfter: DefaultRepeatAfter,
+	}, time.Now().UTC())
+}
+
+func (s *State) addWarningFull(w Warning, t time.Time) {
+	s.writing()
+
+	if s.warnings[w.message] == nil {
+		w.firstSeen = t
+		s.warnings[w.message] = &w
+	}
+	s.warnings[w.message].lastSeen = t
+}
+
+// DeleteOldWarnings delets warnings that up for deletion.
+func (s *State) DeleteOldWarnings() int {
+	s.reading()
+
+	now := time.Now().UTC()
+
+	n := 0
+	for k, w := range s.warnings {
+		if w.IsDeletable(now) {
+			delete(s.warnings, k)
+			s.modified = true
```

Comment:
I don't think we do this in other places,  Prune for example calls  writing in a similar situation

Reply:
ack, will change